### PR TITLE
[Driver][SYCL][FPGA] Improve AOCX archive behaviors with -fsycl-link=image

### PR DESCRIPTION
**Pull request by mdtoguchi**
Use of -fsycl-link=image allows for FPGA users to create a full device binary and store it in an archive to be used later.  This archive can then be picked up and linked in.  The information stored in the archive was not sufficient, causing the device binary to not be 'visible' when linked in.

Improve the situation here by saving all of the information provided by the wrapping step (including symbols) in the archive, but also be sure that this entire final binary is saved as it would be if we were performing a full E2E compilation.

To accomodate the ability to pass around the needed symbols and properties that are used when wrapping the final binaries for any AOCR or AOCX type archives, update the way this information is passed around. Multiple binaries are added to the archive, including the full wrapped .bc file that is gathered and passed along to any subsequent wrapping sequence that occurs.

### CODE REVIEW

**Comment by mdtoguchi** [comment]
dependent on https://github.com/intel/llvm/pull/10850

**Comment by mdtoguchi** [comment]
Also dependent on https://github.com/intel/llvm/pull/11034

**Comment by LU-JOHN** [comment]
When we generate a new wrapped bc file with --sym-prop-bc-files, it is important that the order of the images provided through --sym-prop-bc-files matches the order of the images that is provided through the table.    Otherwise the symbols and properties will be applied to mismatched images.  Do the testcases check this?

Do we have E2E tests for this new functionality?

**Comment by LU-JOHN** [comment]
> We should be able to test this via E2E with emulation. I don't think we have anything in place that performs AOCX and AOCR behaviors on our side.

Will E2E tests be added to this PR or will they be done separately?

**Comment by mdtoguchi** [comment]
> 
> Will E2E tests be added to this PR or will they be done separately?

I think I can add here, add to the tests-e2e area.

**Review by LU-JOHN — COMMENTED** [review]
Testing LGTM.

**Review by hchilama — APPROVED** [review]
LGTM

**Comment by mdtoguchi** [comment]
@asudarsa, please take a look - thanks!

File: sycl/test-e2e/AOT/fpga-aoc-archive.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Test compiler behaviors for -fintelfpga with -fsycl-link.
```
Comment (asudarsa) [inline sycl/test-e2e/AOT/fpga-aoc-archive.cpp:1]: Do we really need this test? It seems to be just a subset of the 'severe' test?

File: clang/lib/Driver/ToolChains/Gnu.cpp [hunk 2]
```diff
@@ -603,10 +603,17 @@ void tools::gnutools::Linker::ConstructJob(Compilation &C, const JobAction &JA,
     // Go through the Inputs to the link.  When a listfile is encountered, we
     // know it is an unbundled generated list.
     for (const auto &II : Inputs) {
-      if (II.getType() == types::TY_Tempfilelist) {
+      // TODO: Incoming file from the unbundling of the AOCX archive is
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Gnu.cpp:606]: I do not understand this TODO comment

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Gnu.cpp:606]: Ideally the incoming file type should be represented as a list, and not an object.  This needs to be improved upon, but it requires more manipulation of the file as the aocx file is passed around.  I can clarify the comment.

File: sycl/test-e2e/AOT/fpga-aoc-archive.cpp (hunk 1 above)
Comment (mdtoguchi) [reply sycl/test-e2e/AOT/fpga-aoc-archive.cpp:1]: That's true.  The severe case covers what this does as well.  I'll clean up and leave the severe case.

**Comment by mdtoguchi** [comment]
@asudarsa, could you take another look?  Thanks!

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 3]
```diff
@@ -9427,17 +9448,23 @@ void OffloadBundler::ConstructJobMultipleOutputs(
...
     // Override type with AOCX/AOCR which will unbundle to a list containing
-    // binaries with the appropriate file extension (.aocx/.aocr).
+    // binaries with the appropriate extension (.aocx/.aocr)
     // TODO - representation of the output file from the unbundle for these
     // types (aocx/aocr) are always list files.  We should represent this
     // better in the output extension and type for improved understanding
     // of file contents and debuggability.
     TypeArg = (InputType == types::TY_FPGA_AOCX) ? "aocx" : "aocr";
-    if (!getToolChain().getTriple().isSPIR())
+    // When the output is a Tempfilelist, we know we are generating unbundling
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:9464]:
```suggestion
    // When the output is a Tempfilelist, we know we are  unbundling
```

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 4]
```diff
@@ -9486,17 +9513,30 @@ void OffloadBundler::ConstructJobMultipleOutputs(
     // FPGA device triples are 'transformed' for the bundler when creating
     // aocx or aocr type bundles.  Also, we only do a specific target
     // unbundling, skipping the host side or device side.
-    if (types::isFPGA(InputType)) {
+    if (types::isFPGA(InputType) || InputType == types::TY_Tempfilelist) {
       if (getToolChain().getTriple().isSPIR()) {
         if (Dep.DependentToolChain->getTriple().getSubArch() ==
             llvm::Triple::SPIRSubArch_fpga) {
+          // mtoguchi
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:9520]: What is this comment for?

File: clang/lib/Driver/ToolChains/Clang.cpp [hunk 5]
```diff
@@ -9725,6 +9782,18 @@ void OffloadWrapper::ConstructJob(Compilation &C, const JobAction &JA,
     WrapperArgs.push_back(
         C.getArgs().MakeArgString(Twine("-kind=") + Twine(Kind)));
 
+    // For FPGA toolchains, we can add additional wrapped bc input files to
+    // the wrapped step.  This is done for AOCR based files that contain the
+    // Symbols and Properties from a previous compilation step.
```
Comment (asudarsa) [inline clang/lib/Driver/ToolChains/Clang.cpp:9787]:
```suggestion
    // For FPGA toolchains, we can provide previously wrapped bc input files to
    // the wrapper step.  This is done for AOCR based files  that will need
    // Symbols and Properties from a previous compilation step.
```

**Review by asudarsa — APPROVED** [review]
Overall changes look good to me. Minor nits. @LU-JOHN will need to approve as well.

Thanks

File: clang/lib/Driver/ToolChains/Clang.cpp (hunk 4 above)
Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/Clang.cpp:9520]: oops, I use my idsid as a marker when editing :)

**Review by asudarsa — APPROVED** [review]

**Comment by mdtoguchi** [comment]
@intel/llvm-gatekeepers, can this be merged?

**Pull request merged**